### PR TITLE
Integrating Airflow with Travis-CI

### PR DESCRIPTION
**Pull request by mistercrunch**
@bolkedebruin, I'm bringing this over as a branch to the main repo to collaborate on it. It's clearly overdue to get better CI for Airflow.

### CODE REVIEW

**Comment by mistercrunch** [comment]
@artwr 
https://travis-ci.org/airbnb/airflow

**Comment by bolkedebruin** [comment]
@mistercrunch excellent! 

Note, it isn't working yet and I am experiencing timeout issues on travis. My (own) focus is also getting a mini-kdc in to ensure security integration testing, I am getting close but not there yet. Most of what I have done is some porting of snakebite's testing, however some stuff is still missing (minicluster for example). I'll be on off working on this when I can find the time.

**Comment by artwr** [comment]
I was having the same issues with https://github.com/artwr/airflow/tree/travis-ci-setup-test

Travis tends to timeout, especially when nothing is output to stdout for a while...

**Comment by mistercrunch** [comment]
If the issue is only stdout, we could use a hack to run commands as subprocesses and periodically issue sometime to stdout. Preferably line by line ascii art :)

**Comment by mistercrunch** [comment]
Looks like there's a standard solution here:
http://docs.travis-ci.com/user/build-timeouts/#Build-times-out-because-no-output-was-received

**Comment by bolkedebruin** [comment]
I am actually using that, but the dependencies take really a lot of time to get installed and it does it several times, eg. for each hadoop type and eventually each python type. We could look at a way to cache this (http://docs.travis-ci.com/user/caching/) or something like this: https://gist.github.com/jsuereth/56bd6f63138f5877f9b8

**Comment by bolkedebruin** [comment]
But first get it running successfully I would say. Without tox this is quite easy, except of course that a lot of the tests fail and tox adds the benefit of being able to run multiple types of testings (ie. cdh/hdp, etc)

**Comment by artwr** [comment]
Even if we do not run the integration testing with hadoop for now, we could use tox to test py2/3 compatibility with requirements and mysql. It would be a huge win.

**Comment by jlowin** [comment]
@mistercrunch could you please add a Python 3.4 (or 3.5 :) ) build? That would be extremely helpful for maintaining compatibility.

**Comment by mistercrunch** [comment]
@jlowin : totally. 3.4 or 3.5? Initially I don't want to put too much stress on travis, so let's just pick one.

**Comment by jlowin** [comment]
I'd vote for starting out cutting edge with 3.5, especially since us Py3-ers seem few and far between :).

**Comment by bolkedebruin** [comment]
So to cut it up it would look something like:
1. MySQL + Postgres
2. Hadoop
   2a. Hadoop Minicluster
   2b. Hive Mini
3. Secured Hadoop

Make sense?

**Comment by bolkedebruin** [comment]
What I am doing (my own branch) to speedup the testing is to use the caching of directories by travis and to use wheels of the packages. I am struggling a little bit with the tox integration as "python setup.py test" does not pick up the wheels. Might need to invoke "tox" directly.

**Comment by bolkedebruin** [comment]
@mistercrunch @artwr I have a reasonable test environment running now (my own branch). 

I have used caching for dependencies by using wheels and by storing them in travis cache. Missing dependencies get added automatically, upgrades of packages not. This speeds up testing from +30mins to around 2mins. Further caching can be done on the hadoop components if required.

Todo: 
- Invalidate cache in case of requirements that need to be upgraded
- The file matching does not entirely work, don't know how to improve currently. Better matching means faster tests
- Selective testing for things that can be tested in travis
- Have an initial mysql / postgres db

Please have a look. The commits are messy at the moment due to a lot of trial and error (geez!)

**Comment by mistercrunch** [comment]
This is awesome! I'll pull your branch in again and see what I can do.

**Comment by mistercrunch** [comment]
Ok this PR is now based off of your current branch, I'm keeping an eye on travis. I'm hoping to find time and take on the selective test part and setting up an initial db.

It looked like deleting the cache can be done manually in the Travis-UI, I think this may be sufficient.

**Comment by bolkedebruin** [comment]
You might have pulled my branch just at the minute it is not working :-). If you did, I'll fix it tomorrow. I have probably a better way of using the wheels anyway.

**Comment by mistercrunch** [comment]
Ok, I'll hold off. It looks like you've done tons of great work though. This is going a game changer!

BTW I'm sure you've figured out that not all the tests work out of the box... Once you're ready I'll jump in and add the missing pieces.

**Comment by bolkedebruin** [comment]
@mistercrunch branch is done. It now fully caches python dependencies after first build. First build can take around 30min, but subsequent builds only around 3min. Further optimization can be done by adding the hadoop downloads to the cache or, only for now, commenting them out as they are not used yet.

I have also included python 3.5 to the build matrix as I saw that was a request somewhere. It fails on the wheel now at the moment as librrabitmq is in the requirements and that is not python 3 compatible yet. Maybe differ between different python versions for the requirements? Anyways its means that 4 builds are now triggered.

It would be nice if airflow pre-fills its connections for testing. In travis you can use mysql (username root password blank or username travis password blank) and postgres (user: postgres). Both have been pre-configured with a "airflow" database that is empty.
See http://docs.travis-ci.com/user/database-setup/ for more details.

Enjoy fixing the tests ;-)!
- Some issues that might popup: in tox.ini you can specify "deps=-rrequirements.txt" but this didn't do anything for me, so I used the commands. Seems to work fine.

**Comment by bolkedebruin** [comment]
@mistercrunch I have update my branch again and fixed a couple of things:
- Copying the airflow_travis.cfg also in my environment :-)
- Instead of using pymysql I am suggesting to use mysqlclient which is the updated MySQL-Python, but with python3 support and actively maintained. It's a drop in replacement

Now integration tests with mysql pass and no errors on sqlalchemy anymore.

**Comment by bolkedebruin** [comment]
Btw I also suggest going down to python 3.4 as a target. Many libs are not at 3.5 making us tracking down a lot of those issues (ie. mysqlclient goes up to 3.4)

**Comment by jlowin** [comment]
@bolkedebruin probably right about 3.4. Got over eager...

**Comment by mistercrunch** [comment]
@bolkedebruin , looks like we're getting close! I just merged your commits here and I'll try to fix the remaining failing tests today.

**Comment by bolkedebruin** [comment]
@mistercrunch awesome! Full python 3 might be a bit off though (sasl not being updated really which is a dependency of pyhs2). I am working on a MiniHiveCluster I might have on ready for testing at the end of the week. But I wouldn't put off merging to master for it.

**Comment by mistercrunch** [comment]
Historical moment:
screen shot 2015-10-21 at 10 39 23 am

Ok, I did disable all the Hive/Presto/Data transfer tests, but it's still a great achievement. Props to  @bolkedebruin who did all the hard work !

**Comment by mistercrunch** [comment]
A'ight I rebased, I think this is good to go.

**Comment by bolkedebruin** [comment]
Nice! It certainly looks that way. Maybe you want to include: http://docs.travis-ci.com/user/status-images/ ?

File: requirements.txt [hunk 1]
```diff
@@ -16,29 +17,28 @@ gunicorn
...
+# snakebite -> not compatible with Python 3
 sphinx
 sphinx-argparse
 sphinx_rtd_theme
 Sphinx-PyPI-upload
 sqlalchemy
 thrift
 jaydebeapi
-#pymssql
+mysqlclient
```
Comment (ldct) [inline requirements.txt:42]: this seems to introduce an install-time dependency on mysql as mysqlclient looks for mysql_config

**Pull request merged**